specs/version_format_specification.md

```diff
@@ -20,6 +20,7 @@ A version string must use the following character set:
 - Minus (`-`) separates the version and release parts.
 - Tilde (`~`) is a prefix that always compares lower.
 - Caret (`^`) is a prefix that always compares higher.
+- Underscore (`_`) separates the version from the name of an artifact, if both are included in the same string.
```

Comment:
I am aware, but the problem is that this is already factually the case for DDIs that we support, so we can't go back on that. We use `-` for prefixes within the names, and `_` to separate the name with the version. I know that this follows the Debian style rather than the RPM style, but I think that ship has sailed? This is not meant to impose a new format on packages (at least until we release systemd-packaged ;-) ), but just for the stuff that we do here (IE: DDIs, UKIs) so hopefully it should be ok?

Reply:
@keszybz I could explicitly rename/reword this to include DDI/UKI in the title to ensure we don't get confused with packaging rules if that helps?

Reply:
instead of declaring "_" a separator I'd just say "other characters, including '_', ':', ';', ',', '+' are not considered part of version strings compatible with this specification, and may hence can be used as separators preceeding or following a version string.

Reply:
But again the problem is that we already use `_` in existing, released code for DDIs, and the purpose of this is to define how the version applies to DDIs, so I don't see how to get away without mentioning this explicitly?

Reply:
> instead of declaring `_"` a separator I'd just say "other characters, including `_`, `:`, `;`, `,`, `+` are not considered part of version strings compatible with this specification, and may hence can be used as separators preceeding or following a version string.

Sorry, but I don't think this works at all. As I wrote above, `_` is used in significant majority of strings that we would compare using this logic when using stuff that is packaged with rpm. If we try to disallow `_`, I think we'd not only break compatibility, but we're also unlikely to ever get any kind of compliance with the spec again, because suddenly almost everything would be in violation. E.g. every `bootctl list` and so on on Fedora.

The other characters are less problematic, because they are much less widely used.

As discussed before, `+` is used in semver. And semver is steadily gaining popularity. Even if we can't get rid of `+` used for boot counting, we shouldn't at least make the situation worse.

Similarly, `:` is used to seperate the epoch in full rpm version strings, but the epoch is not written to file names, so most of the time we wouldn't see it in comparisons. Right now we handle `:` in a reasonable way, as a separator. If both strings have an epoch, then it is compared numerically, and the right result occurs. rpmdev-vercmp has special support for skipping the epoch (`'0:253' == '253'`), which we obviously don't implement, but that's OK.

AFAIK, `;` and `,` are not used by anything. `;` is special for the shell, and `,` also is in some cases. If EFI firmwares support it, I think it would be reasonable to start using either of those two as the boot count separator.

Reply:
> rpm, dpkg and semver are already not compatible:

yep, precisely. All I am saying it would be supernice to have this summary somewhere in the spec, to clarify things. in particular incompatibilities between the four specs would make sense to mention, as well as the "common core" of all four specs, and finally the "compressed charsets" of each spec.

but this is just wishful thinking on my side, nice stuff to have, to position this spec sensibly in relationship to the other specs, that's all. would be great to have as part of the "creation" of this new spec, but I totally can understand if you don't want to bother with this now (though in a way your comment here already covers 80% of what I am looking for, and hence cut just be copy/pasted into the spec), or ever.

Reply:
Yes, a section about incompatibilities would be nice-to-have.

Reply:
> but then in the DDI and BLS specs pick a specific one and forbid the other

Pfffff. That's just terrible. DDI is new, can't we just adjust DDI version comparison to behave as BLS which has been around for 8 years?

Reply:
That's a bit of a bummer. I hope we can figure out a way to modify the DDI naming in a way that maintains some sort of backwards-compatibility but reconciles the name-version handling with the BLS and its version spec.

Or if that can't be done, just say in the DDI spec how the splitting should be done, and than that the version part is compared using the version spec.

Reply:
@bluca which location specifically do you have in mind where we already rely on "_" as separator? portabled? or anywhere else?

Reply:
portabled and sysupdated at least, not sure if it's wired in dissect itself yet https://www.freedesktop.org/software/systemd/man/sysupdate.d.html

Reply:
Update to avoid mention the topic and leave it unspecified, and changed the individual specs to specify the separator instead, would this work?